### PR TITLE
refactor(packagist): Refactor schema

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Reorganize schema before adding de-minify feature

## Context

- Ref: #19863

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/packagist/schema.ts [hunk 1]
```diff
@@ -1,28 +1,34 @@
...
-export const ComposerReleaseArray = z
-  .array(ComposerRelease.nullable().catch(null))
-  .transform((xs) =>
-    xs.filter((x): x is z.infer<typeof ComposerRelease> => x !== null)
+const Version = z
+  .string()
+  .refine((v) => versioning.isSingleVersion(v), 'Invalid version');
+
+export const ComposerRelease = z
+  .object({
```
Comment (zharinov) [inline lib/modules/datasource/packagist/schema.ts:11]: Strict part

File: lib/modules/datasource/packagist/schema.ts [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
-      url: z.string(),
-    })
-    .transform((x) => x.url)
-    .nullable()
-    .catch(null),
-  time: z.string().nullable().catch(null),
-});
+const Version = z
+  .string()
+  .refine((v) => versioning.isSingleVersion(v), 'Invalid version');
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:8]: Maybe this mistake was made earlier but we should not be using any `versioning` within `lib/modules/datasource/*/*`. Versioning filtering should be done by the datasource index/dispatcher and not hardcoded into datasource implementations

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/packagist/schema.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/datasource/packagist/schema.ts:8]: Example search showing that this is unique:

image

Comment (zharinov) [reply lib/modules/datasource/packagist/schema.ts:8]: But we still need to use version comparison and validity check for deciding on most fresh `homepage` and `sourceUrl` (except for unstable releases like `dev-main`)?

Comment (zharinov) [reply lib/modules/datasource/packagist/schema.ts:8]: Actually, I think we can safely assume the first release to be the latest one

Comment (rarkins) [reply lib/modules/datasource/packagist/schema.ts:8]: Yes, use the first

**Review by rarkins — APPROVED** [review]

File: lib/modules/datasource/packagist/schema.ts [hunk 3]
```diff
@@ -71,10 +70,12 @@ export function parsePackagesResponses(
 
       releases.push(dep);
 
-      if (!maxVersion || versioning.isGreaterThan(version, maxVersion)) {
```
Comment (viceice) [inline lib/modules/datasource/packagist/schema.ts:74]: so this is a partial revert?

**Review by viceice — APPROVED** [review]

File: lib/modules/datasource/packagist/schema.ts (hunk 3 above)
Comment (zharinov) [reply lib/modules/datasource/packagist/schema.ts:74]: Yes

**Pull request merged**